### PR TITLE
Add Tensor product

### PR DESCRIPTION
**Pull request by nahumsa**
This adds the [tensor product of graphs](https://en.wikipedia.org/wiki/Tensor_product_of_graphs) for `PyDiGraph` and `PyGraph`.

This PR is related to #440.

### CODE REVIEW

File: releasenotes/notes/add-tensor-product-651db371a96d650c.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+---
+features:
+  - |
+    Add two new functions which calculates the tensor product of two graphs :func:`~retworkx.graph_tensor_product`
+    for undirected graphs and :func:`~retworkx.digraph_tensor_product` for directed graphs. For example:
+    
+    .. jupyter-execute::
+      import retworkx
```
Comment (mtreinish) [inline releasenotes/notes/add-tensor-product-651db371a96d650c.yaml:8]:
```suggestion
    .. jupyter-execute::

      import retworkx
```

**Review by mtreinish — COMMENTED** [review]
Thanks for doing this. It LGTM, just a couple inline questions/suggestions but other than that I think this is mostly ready.

File: src/tensor_product.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+fn tensor_product<Ty: EdgeType>(
+    py: Python,
+    first: &StablePyGraph<Ty>,
+    second: &StablePyGraph<Ty>,
+    undirected: bool,
+) -> (StablePyGraph<Ty>, ProductNodeMap) {
+    let mut final_graph = StablePyGraph::<Ty>::default();
+
+    let mut hash_nodes = HashMap::with_capacity(first.node_count() * second.node_count());
```
[the remark this answers is not in the document]
Comment (nahumsa) [reply src/tensor_product.rs:32]: Sure! I'll double-check the number of edges to use `with_capacity`.

File: src/tensor_product.rs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+    let nodes_first = first.node_indices();
+    let nodes_second = second.node_indices();
+
+    let cross = nodes_first.flat_map(|x| nodes_second.clone().map(move |y| (x, y)));
```
[the remark this answers is not in the document]
Comment (nahumsa) [reply src/tensor_product.rs:37]: Yep, this is more readable code. I'm not sure where I was going with iterator inside the iterator, but I'll rewrite using a loop.

[the remark this answers is not in the document]
Comment (nahumsa) [reply src/tensor_product.rs]: Added in [`444fa48`](https://github.com/Qiskit/retworkx/pull/472/commits/444fa4805066d09cdfa23e696de108c17b46e530)

File: tests/digraph/test_tensor_product.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import unittest
+import retworkx
+
+
+class TestTensorProduct(unittest.TestCase):
+    def test_directed_null_tensor_null(self):
+        graph_1 = retworkx.PyDiGraph()
+        graph_2 = retworkx.PyDiGraph()
+
+        graph_product, _ = retworkx.digraph_tensor_product(graph_1, graph_2)
```
Comment (mtreinish) [inline tests/digraph/test_tensor_product.py:22]: It might be good to check the output `ProductNodeMap` is correct too.

**Review by mtreinish — COMMENTED** [review]
I think the rust code lgtm now, thanks for the quick update. I think we need to make some improvements to the tests though (sorry I only skimmed them in my earlier review). The comments I left inline apply to pretty much all the tests though.

File: tests/digraph/test_tensor_product.py (hunk 4 above)
Comment (nahumsa) [reply tests/digraph/test_tensor_product.py:22]: Added in [`c2795ad`](https://github.com/Qiskit/retworkx/pull/472/commits/c2795ade2c1278430503f2757a58891ec9c326b0)

File: tests/digraph/test_tensor_product.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        self.assertEqual(graph_product.num_nodes(), 0)
+        self.assertEqual(graph_product.num_edges(), 0)
+
+    def test_directed_path_2_tensor_path_2(self):
+        graph_1 = retworkx.generators.directed_path_graph(2)
+        graph_2 = retworkx.generators.directed_path_graph(2)
+
+        graph_product, _ = retworkx.digraph_tensor_product(graph_1, graph_2)
+        self.assertEqual(graph_product.num_nodes(), 4)
+        self.assertEqual(graph_product.num_edges(), 1)
```
[the remark this answers is not in the document]
Comment (nahumsa) [reply tests/digraph/test_tensor_product.py:36]: Added in [`c2795ad`](https://github.com/Qiskit/retworkx/pull/472/commits/c2795ade2c1278430503f2757a58891ec9c326b0)

File: tests/digraph/test_tensor_product.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        self.assertEqual(graph_product.num_nodes(), 0)
+        self.assertEqual(graph_product.num_edges(), 0)
+
+    def test_directed_path_2_tensor_path_2(self):
+        graph_1 = retworkx.generators.directed_path_graph(2)
+        graph_2 = retworkx.generators.directed_path_graph(2)
+
+        graph_product, node_map = retworkx.digraph_tensor_product(graph_1, graph_2)
+        expected_node_map = {(0, 0): 0, (0, 1): 1, (1, 0): 2, (1, 1): 3}
+        self.assertTrue(node_map == expected_node_map)
```
Comment (mtreinish) [inline tests/digraph/test_tensor_product.py:32]:
```suggestion
        self.assertEqual(node_map, expected_node_map)
```

File: tests/graph/test_tensor_product.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        self.assertEqual(graph_product.num_edges(), 0)
+
+    def test_path_2_tensor_path_2(self):
+        graph_1 = retworkx.generators.path_graph(2)
+        graph_2 = retworkx.generators.path_graph(2)
+
+        graph_product, node_map = retworkx.graph_tensor_product(graph_1, graph_2)
+
+        expected_node_map = {(0, 0): 0, (0, 1): 1, (1, 0): 2, (1, 1): 3}
+        self.assertTrue(node_map == expected_node_map)
```
Comment (mtreinish) [inline tests/graph/test_tensor_product.py:33]:
```suggestion
        self.assertEqual(node_map, expected_node_map)
```

**Review by mtreinish — APPROVED** [review]
LGTM, thanks for updating the tests

**Pull request merged**